### PR TITLE
Re-implement config_option using get_as

### LINKED ISSUE #1113: Option parsing for string arguments beginning with quotes is broken
Consider the following option that takes a `std::string` as a query in some hypothetical query language:

```sh
# broken, returns pec::trailing_characters
my_app '--query="text" in field'

# works as expected
my_app '--query=field contains "text"'

# also works:
my_app '--query="\"text\" in field"'
```

The reason being that CAF's string parser is used to parse the option, and when the whitespace-stripped argument begins with either a single or a double quote, CAF only parses up until the matching quote, and returns a `pec::trailing_characters` error if further characters are available. This behavior is quite unintuitive.

### PR DESCRIPTION
**Pull request by Neverlord**
Using `get_as` for implementing `config_option` gives us synchronization with any type for free. We can also remove the custom CLI parsing, since the new `sync` function automatically lifts strings as needed. The `get_as` utility also knows to leave strings untouched for string arguments, which fixes #1113 as a neat side effect.

### CODE REVIEW

**Review by josephnoir — COMMENTED** [review]
It looks like the PR removes quite a few test cases. Although that might just have been my impression.

File: libcaf_core/caf/make_config_option.hpp [hunk 1]
```diff
@@ -22,70 +22,48 @@
...
-CAF_CORE_EXPORT expected<config_value>
-parse_impl(std::string* ptr, string_view str);
-
 template <class T>
-expected<config_value> parse_impl_delegate(void* ptr, string_view str) {
-  return parse_impl(reinterpret_cast<T*>(ptr), str);
+config_value get_impl(const void* ptr) {
+  config_value result;
+  auto err = result.assign(*static_cast<const T*>(ptr));
+  static_cast<void>(err); // Safe to discard because sync() fails otherwise.
```
[the remark this answers is not in the document]
Comment (Neverlord) [reply libcaf_core/caf/make_config_option.hpp:53]: Not really, the comment only points out that we know that `assign` will succeed here. If it wouldn't, `sync` would also break which then reports the error at the point of assigning an "unconvertible" value.

File: libcaf_core/test/config_option.cpp [hunk 2]
```diff
@@ -32,6 +34,148 @@ using std::string;
...
+      CAF_FAIL("failed to parse result settings: " << parsed.error()
+                                                   << "\nINPUT:\n"
+                                                   << res << '\n');
+  }
+
+  template <class Res>
+  void add_test(std::vector<std::string> cli, std::string conf, Res&& res) {
+    return add_test(std::move(cli), std::move(conf), std::forward<Res>(res),
+                    [](const state&) { return true; });
+  }
```
[the remark this answers is not in the document]
Comment (Neverlord) [reply libcaf_core/test/config_option.cpp:137]: The baselines test the full roundtrip from parsing CLI arguments and config "file" content to printing the content of the generated configuration. We only use the predicate to check whether ref-syncing also works, but we don't need to repeat this at every test. Also, some parameters have no ref-syncing to begin with, so there's no need for an extra predicate.

File: libcaf_core/test/config_option.cpp [hunk 3]
```diff
@@ -32,6 +34,148 @@ using std::string;
...
+
+BEGIN_FIXTURE_SCOPE(fixture)
+
+SCENARIO("options on the CLI override config files that override defaults") {
+  for (size_t index = 0; index < baselines.size(); ++index) {
+    state st;
+    st.run(baselines[index], index);
+  }
+}
+
```
[the remark this answers is not in the document]
Comment (Neverlord) [reply libcaf_core/test/config_option.cpp:178]: > Do you think it a good idea to mix test ... versions in the same file

The important thing is *having* tests. I don't think dogmatically re-writing all tests in BDD-style would be worth the time.

> And the fixture is only needed for the scenario above, correct?

Technically, yes. But I think having the fixtures scope ending somewhere in the middle would only make things more complicated than it ought to be. We don't have a rule again extending / using the fixture in the other tests eventually. 🙂

File: libcaf_test/caf/test/unit_test.hpp [hunk 4]
```diff
@@ -642,32 +642,81 @@ using caf_test_case_auto_fixture = caf::test::dummy_fixture;
...
 // -- CAF_CHECK* predicate family ----------------------------------------------
 
-#define CAF_CHECK_EQUAL(x, y) CAF_CHECK_FUNC(::caf::test::equal_to, x, y)
+#define CAF_CHECK_EQUAL(x_expr, y_expr)                                        \
+  ([](const auto& x_val, const auto& y_val) {                                  \
+    return ::caf::test::detail::check_bin(x_val == y_val, __FILE__, __LINE__,  \
+                                          #x_expr " == " #y_expr,              \
+                                          caf::deep_to_string(x_val),          \
+                                          caf::deep_to_string(y_val));         \
+  })(x_expr, y_expr)
```
[the remark this answers is not in the document]
Comment (Neverlord) [reply libcaf_test/caf/test/unit_test.hpp:651]: Technically, these aren't necessary. I don't remember why I've put them there, probably just habit of surrounding everything in macros with parentheses. Removed.

File: libcaf_core/test/config_option.cpp [hunk 5]
```diff
@@ -32,6 +34,148 @@ using std::string;
...
+  state() {
+    config_option_adder{options, "?my.app"}
+      .add(my_app_s1, "s1", "")
+      .add(my_app_vector, "vector,v", "")
+      .add(my_app_severity, "severity,s", "")
+      .add(my_app_request, "request,r", "")
+      .add(my_app_request_pair, "request-pair,R", "");
+    config_option_adder{options, "sys"}
+      .add<std::string>("path,p", "")
+      .add<int8_t>("threads,tTd", "");
```
[the remark this answers is not in the document]
Comment (Neverlord) [reply libcaf_core/test/config_option.cpp]: Oops, I've forgot to add the baselines for these options. Done.

**Comment by Neverlord** [comment]
> It looks like the PR removes quite a few test cases.

Yes, but it also removes quite a lot of code. 🙂

This PR is a net reduction in code. The removed tests became either redundant or obsolete.

Thanks for the review! I've addressed all individual comments in the latest commit.

**Comment by josephnoir** [comment]
You are welcome! Looks good :)

**Pull request merged**